helper/awsutil/generate_credentials.go

```diff
@@ -54,6 +59,20 @@ func (c *CredentialsConfig) GenerateCredentialChain() (*credentials.Credentials,
 			"static AWS client credentials haven't been properly configured (the access key or secret key were provided but not both)")
 	}
 
+	roleARN := os.Getenv("AWS_ROLE_ARN")
```

Comment:
Does the AWS SDK handle any of these sorts of environment variables under the hood? So that we could delegate this to them?

Reply:
It does handle them but only when you keep the default credentials chain.

As Vault is creating its own chain, we need to create a `WebIdentityProvider` and add it to the chain, so it's not something that the SDK will do by itself, that's one of the reasons that I had to handle the env variables myself instead of delegating it.

Reply:
I don't think this is correct.

https://github.com/aws/aws-sdk-go/blob/v1.26.8/aws/defaults/defaults.go#L93 and https://github.com/aws/aws-sdk-go/blob/v1.26.8/aws/defaults/defaults.go#L107 shows that the default cred chain consists of

```go
	[]credentials.Provider{
		&credentials.EnvProvider{},
		&credentials.SharedCredentialsProvider{Filename: "", Profile: ""},
		RemoteCredProvider(*cfg, handlers),
	}
```
All three of these are added by us in the same order. As this includes EnvProvider and that checks AWS_ROLE_SESSION_NAME it shouldn't be necessary to manaully check env vars.

Reply:
Another way to approach this might be, if wanting a WebIdentityRoleProvider is a general-purpose thing, suggest in the aws-sdk-go repo to simply make it part of the default set.

Reply:
I've been looking into this more carefully. I can see some different behaviour providing your own credentials chain vs using the default one:

* Using the default credentials chains, the sdk detects the WebIdentityProvider Env variables correctly and internally it calls `NewWebIdentityRoleProvider`.
This means that calling any AWS service without creation the custom credentials chain makes IRSA work by default.

* When using the custom credentials chain with the call to `NewWebIdentityRoleProvider` commented, the `WebIdentityProvider` configuration gets overridden by the provided chain providers.

This means that we in fact need to provide the `WebIdentityProvider` if we are to create a custom credentials chain because, by calling `Retrieve` on the other providers, it will never call the WebIdentity code.